PerimeterXModule/Internals/ReverseProxy.cs

```diff
@@ -0,0 +1,240 @@
+				{
+					Debug.WriteLine("ReverseProxy responeded with none 200 status", PxConstants.LOG_CATEGORY);
+					Debug.WriteLineIf(!(response == null), "Response status {0}", PxConstants.LOG_CATEGORY);
+					return false;
+				}
+
+				byte[] responseData = server.GetResponseStreamBytes(response);
+
+				// Send the response to client
+				context.Response.Headers.Add(response.Headers);
+				context.Response.ContentEncoding = Encoding.UTF8;
+				context.Response.ContentType = response.ContentType;
+				context.Response.OutputStream.Write(responseData, 0, responseData.Length);
+
+				// Handle cookies to navigator
+				server.SetContextCookies(response);
+
+				// Close streams
+				response.Close();
+				return true;
+			}
+			catch (Exception e)
+			{
+				Debug.WriteLine("Unexpected error while processing reverse request: " + e.Message, PxConstants.LOG_CATEGORY);
+				return false;
+			}
+
+		}
+		/**
+		 * <summary>
+		 * Forward requests for PerimeterX client
+		 * </summary>
+		 * <param name="context">HttpContext</param>
+		 */
+		public void ReversePxClient(HttpContext context)
+		{
+			Debug.WriteLine("Fetching Client", PxConstants.LOG_CATEGORY);
+			string contentType = CONTENT_TYPE_JAVASCRIPT;
+			string defaultResponse = DEFAULT_CLIENT_VALUE;
+			if (!PxConfig.FirstPartyEnabled)
+			{
+				Debug.WriteLine("First party is disabled, rendering default response", PxConstants.LOG_CATEGORY);
+				RenderPredefinedResponse(context, contentType, defaultResponse);
+			}
+			string uri = string.Format("/{0}/main.min.js", PxConfig.AppId);
+			bool success = ProcessRequest(context, PxConfig.ClientHostUrl, uri);
+
+			if (!success)
+			{
+				Debug.WriteLine("Redirect XHR returned bad status, rendering default response", PxConstants.LOG_CATEGORY);
+				RenderPredefinedResponse(context, contentType, defaultResponse);
+			}
+
+		}
+
+		/**
+		 * <summary>
+		 * Forward any sensor activities back to PerimeterX servers
+		 * </summary>
+		 * <param name="context">HttpContext</param>
+		 */
+		public void ReversePxXhr(HttpContext context)
+		{
+			string defaultResponse;
+			string contentType;
+
+			if (context.Request.Url.AbsolutePath.EndsWith(".gif"))
+			{
+				defaultResponse = Encoding.Default.GetString(DEFAULT_EMPTY_GIF_VALUE);
+				contentType = CONTENT_TYPE_GIF;
+			}
+			else
+			{
+				defaultResponse = DEFAULT_JSON_VALUE;
+				contentType = CONTENT_TYPE_JSON;
+			}
+
+			if (!PxConfig.FirstPartyEnabled || !PxConfig.FirstPartyXhrEnabled)
+			{
+				Debug.WriteLine("First party is disabled, rendering default response", PxConstants.LOG_CATEGORY);
+				RenderPredefinedResponse(context, contentType, defaultResponse);
+			}
+			string uri = context.Request.RawUrl.Replace(XhrReversePrefix, "");
+
+			string vid = null;
+			HttpCookie pxvid = context.Request.Cookies.Get("pxvid");
+			if (pxvid != null)
+			{
+				vid = pxvid.Value;
+			}
+			else if ((pxvid = context.Request.Cookies.Get("_pxvid")) != null)
+			{
+				vid = pxvid.Value;
+			}
```

Comment:
According to [this](https://stackoverflow.com/questions/13058574/check-if-cookie-exists), if cookie with this name doesn't exist, then a new cookie with this name is created and returned? If this is the case, it'd be better to use `context.Request.Cookies.AllKeys.Contains("pxvid")` instead.

Then can do something like:

```
string vid = string.Empty;
if (context.Request.Cookies.AllKeys.Contains("pxvid")) {
	vid = context.Request.Cookies.Get("pxvid").Value;
}
else if (context.Request.Cookies.AllKeys.Contains("_pxvid") {
	vid = context.Request.Cookies.Get("_pxvid").Value;
}
```

Reply:
👎 
Read that doc again
the new cookie will be created on the response only
not on the request